### DISCUSSION TITLE
UI Questions/comments

### DISCUSSION THREAD

**Discussion by josh-griffin** [openmsupply/openmsupply-client#175] [General]
I didn't know where to put this. I'm giving this a go. Hopefully it works

**Comment by josh-griffin** [comment]
### List view row select and navigation.

- I find myself trying to select a row by pressing anywhere on the row out of habit / muscle memory.
- I also have tried to copy + paste some test in a row which has resulted in a navigation to the detail page and frustrating experience just trying to copy a customers name.

Wondering if a column for `navigate and view the details of the invoice` would be good. 
- Have the row select action triggered on a row press
- The more specific cell press trigger the navigation event

Rather than the reverse.

The navigation event causes more frustrating on an accident than a selecting a row, so it should be the action that is harder to accidentally trigger (something with pits of success, or something)

_Could you remove the checkbox, have a row highlight, replace the checkbox with a "navigation column" if number of columns is a concern?_

**Reply by mark-prins** [reply]
I haven't had a problem with it myself but can see where you're coming from.
Agree that the lowest-risk action should be the default, so would be up for adding a column which has a button which navigates to the item.
We could add a double-click as well, for the twitchy fingered users who like that sort of thing.

I'd prefer to keep the checkbox - I wonder though if we could click the row to change the checkbox state, would that be weird? Less fiddly than clicking on a little box

**Reply by mark-prins** [reply]
Another viewpoint: "Feels like icon would be naff on mobile screen but good on desktop?"

**Reply by josh-griffin** [reply]
- No I think keeping the checkbox is good.
- I have seen enough people use mobile on a tablet and have no idea to click on a row and end in randomly clicking places to know that having an indication of something being pressable is good

**Comment by josh-griffin** [comment]
Editable cells in a table. I haven't seen any examples a part from the batch modal - if there are some, can someone direct me to it? Otherwise
- Can we get an example where the cells are editable directly?
- Can the edit modal table and the normal table be the same? Are they different for a reason?

Bonus: Do we want to use numeric extras on the input i.e. +/- buttons

![image](https://user-images.githubusercontent.com/35858975/134826918-8cf081c5-3f8f-4cc1-b128-da33cf3d6850.png)

**Reply by josh-griffin** [reply]
For this and the below question I have since discovered:

`We have chosen that there will be no in-line editing even for quantity.`

Which in hindsight makes sense why there's no examples, I had just assumed that cells in the table were editable!

This seems a bit crazy to me. However I don't really use the app in it's intended use and haven't seen people do it, so what do I know?

**Reply by mark-prins** [reply]
quantity was the field that was most likely to be an editable one ( comment was being handled differently, maybe reference would be? )

the thinking was that we need more than just validation for quantity - what if you want to add more than the available quantity of an item? can you add another line with a different batch? or add a placeholder line? we could allow increase & decrease of an item quantity within the bounds of available stock for that batch - but it wouldn't cover all scenarios and so you'd still need another way to edit the line quantity. felt like it was getting very complicated for little benefit

**Comment by josh-griffin** [comment]
Validation: 
- What is the desired UI behaviour for validation and errors - allow input and show an error? (This is for form inputs and cells)
- Where is the feedback shown?

For example:
- For cells, do we only allow numbers to be entered? We can disallow `.` or `-` for CI inputs, but what is the pattern for prices. For example we need to allow `10.10` - what if the user enters `10.` - do we show an error? Revert to an old value at some point?
- For form inputs, looks like the code and item are editable, they could conceivably not match, where is the error shown? (This modal doesn't have too many examples, but for example a supplier invoice modal would hopefully be the same thing, and we might need validation on a batch name to not be too long)


Edit:

I just noticed there's an allocation error message here:

![image](https://user-images.githubusercontent.com/35858975/134827163-128ffd41-4510-4df0-abaf-79ef02c5a22f.png)

Still curious about for cells (Also think that an error message on the RHS of an input makes the horizontal space for an input quite massive)

**Reply by josh-griffin** [reply]
as per: https://github.com/openmsupply/openmsupply-client/discussions/175#discussioncomment-1387572

Don't need cell validation

Will still mention that the error placement for form inputs seems odd and will end up being inconsistent for forms, probably?

**Reply by mark-prins** [reply]
ah. well. *cough* I know it _looks_ like an error, but the `1500 allocated` was more of an indication.
It is in red though which is why I'm concerned about it now.

we do need input validation and I was just confusing cells and input fields.

**Comment by josh-griffin** [comment]
Are customer invoices meant to be per item rather than per batch?

**Reply by richardmoizeau** [reply]
Ongoing discussion
- Possibility to change the view of the General Tab from "per item" view to "per item with batch data" (we asked our designer to provide a preview for this)
- we don't want another tab to show the batch details - an additional tab should only show different data and not another view of the same data

**Comment by josh-griffin** [comment]


**Reply by richardmoizeau** [reply]
https://docs.google.com/spreadsheets/d/1MfRa6U-QchAbFXU5_moyM5jxJo1wjhhmKfpD4gMR0QE/edit#gid=697166990&range=C10

**Comment by josh-griffin** [comment]
_Creating an invoice flow_

What is the flow for creating an invoice?
Example:
1. Click add shipment
2. Navigate to new shipment page
3. Modal for selecting a customer is open (And can't be closed? Where can you cancel/go back? if you press the backdrop do you navigate back?)
4. Searching for a customer: Is the list just filtered when you type? Filtered on just the name? (There are potentially 1000's of customers, apparently, so need to research if this is feasible if it's the desired behaviour)

Bonus:
How do I change customer once selected / what does the pencil icon do? (Does the behaviour for the customer ref pencil and  customer name pencil differ?)

**Reply by richardmoizeau** [reply]
I have written scenarios in below spreadsheet for some of the activities related to the Outbound Shipments included "Creation of a Ad Hoc Outbound Shipment from the Outbound Shipment list using [New Outbound Shipment] button". Long but I prefer to be clear. 

I haven't worked on the "Change the selected Customer" workflow. I can give it a go

https://docs.google.com/spreadsheets/d/1MfRa6U-QchAbFXU5_moyM5jxJo1wjhhmKfpD4gMR0QE/edit#gid=697166990&range=C6

**Comment by josh-griffin** [comment]
*Add / Edit Item customisations*

- Theres no example of the `Other tab`
- Is this where customisations/enhancements go? For example, if a similar modal will be for dispensing, and a customer wants a `PrescribedDate` customisation, would that be in the `other` tab?

**Reply by mark-prins** [reply]
for customer invoices there's the `Transport` stuff, which we don't have a design for yet, and the pricing stuff, which we don't have a design for yet.

**Comment by josh-griffin** [comment]
_Finalising flow_

Is (what we currently call) finalising the `save & confirm allocation button` (I am unsure as this seems to be a button easy for mistakes)?

Assuming so:
- What happens afterward? Is there a message/dialog? Are you navigated back to the list?
- What happens on failing to confirm?
- What happens during? (Can take ~.5 seconds+ to complete, we need to ensure the user either can't enter data (or do we just roll it back?) - do we 'disable' everything and show a spinner on the button? Show a full-screen 'LOADING' spinner? Something else?

**Reply by richardmoizeau** [reply]
- What happens afterward? Is there a message/dialog? Are you navigated back to the list?

1. Save & Confirm Allocation: user have allocated stock to the shipment and stock is now reserved (cannot be used for another shipment) - Status changes from Draft to Allocated
2. Save & Confirm Picking: user can print a pick slip. Once shipment has been prepared and is ready to be shipped, user can confirm picking. Status changes from Allocated to Picked
3. Save & Confirm Shipment: user confirms that shipment has left the facility and is now in transit. Shipment is now on read-only and status changes from Picked to Shipped. 
4. Status changes from Shipped from Delivered when receiving store confirms the delivery. No action required from the supplying store. 

A dialog window is necessary asking the user if they want to confirm allocation/picking/shipment. In both case user is navigated back to the shipment. 

What happens on failing to confirm?

You mean if the user says "No" to the question: "Are you sure you want to proceed ?" - User stays on the Shipment screen and nothing has changed. We could imagine a notification to indicate that "Allocation has not been confirmed". 

What happens during? (Can take ~.5 seconds+ to complete, we need to ensure the user either can't enter data (or do we just roll it back?) - do we 'disable' everything and show a spinner on the button? Show a full-screen 'LOADING' spinner? Something else?

Good to know this could take a few seconds. I guess you are right, we should disable everything and indicates somehow that the allocation or the shipment is being confirmed.

**Reply by josh-griffin** [reply]
Sorry I forgot about this :)

- How do you change the status? Can you click on this? Then Or does the `save & ..` button always change to say the Next status? So if you want to go from `new` to `shipped` you click 3 times?
![image](https://user-images.githubusercontent.com/35858975/136715479-3afb3eb2-e099-421b-b1cd-cd757a19e483.png)
- How do you print a pick slip? (Also what does a pick slip look like?)
- Is there an indication other than the status at the bottom that the invoice is read only?

**Reply by richardmoizeau** [reply]
Bonjour ! :-)

1st question:

I think the pop-up window that provides status details is for information only. 

For Outbound Shipment: 
- You change the status by clicking on `Save & XXX` - the button changes based on the next status
- Status `Delivered` is automatically updated when the receiving store is acknowledging the delivery on their side
For Inbound Shipment:
- Following status are automatically updated when supplying store has confirmed then: `New > Allocated > Picked > Shipped`
- When current status is `Shipped`, user at the receiving store can `Save & Confirm Delivery`
Hope it is clear... Let me know if not. 

2nd question (pickslip):
- a pick slip is a document to help preparing/picking an order. It should contain following information: item code, item name, batch number, expiry date, quantity, store location. Ideally, the order of the items should be based on the store locations (so the user does not have to run around) 
- below an example of pick slip in mSupply:
<img width="1571" alt="Screen Shot 2021-10-11 at 12 59 23 PM" src="https://user-images.githubusercontent.com/74992958/136717223-11ef24ad-8f90-4397-9bec-27d9b31f8104.png">

3rd question (read-only): 
- no preview so far - happy for you and Mark to give it a go (disabled button, a clear mention that document in on read only....)

**Comment by josh-griffin** [comment]
_Viewing a finalised invoice_

An allocated? finalised? (sorry i'm new here) invoice can't be edited, what do each of the components look like when they are disabled?

**Reply by richardmoizeau** [reply]
Status of the shipment: 
Draft > Allocated > Picked > Shipped > Delivered

On the Supplying Store side, a shipment is no longer editable when it reaches the status `Shipped`. We don't have any preview for disabled buttons. Do you think you could give it a go ?
